sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/CatalystRowConverter.scala

```diff
@@ -126,7 +169,24 @@ private[parquet] class CatalystRowConverter(
 
   // Converters for each field.
   private val fieldConverters: Array[Converter with HasParentContainerUpdater] = {
-    parquetType.getFields.zip(catalystType).zipWithIndex.map {
+    // In case of schema merging, `parquetType` can be a subset of `catalystType`.  We need to pad
+    // those missing fields and create converters for them, although values of these fields are
+    // always null.
```

Comment:
This is a good question. Tested the following snippet against 1.4.1, and it doesn't work as expected:

```
import sqlContext.implicits._
import org.apache.spark.sql.types._

sqlContext
  .range(1).select('id as 'a, 'id as 'b)
  .write.mode("overwrite").parquet("file:///tmp/schema")

sqlContext
  .read
  .schema(StructType(StructField("a", LongType, false) :: Nil))
  .parquet("file:///tmp/schema").show()
```

So at least this won't be a regression. This worths further investigation though.